File: main/src/main/java/com/example/main/data/source/remote/MainRepository.java
package com.example.main.data.source.remote;

import android.support.annotation.NonNull;


import com.example.core.data.model.Resource;
import com.example.core.data.model.movie.Movie;
import com.example.core.util.SingleLiveEvent;

import java.util.Objects;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class MainRepository {

    private WebServiceMain webService = new ApiUtilsMain().getListMovieService();

    public void getMovie(SingleLiveEvent<Resource<Movie>> movieSingleLiveEvent, Integer page) {
        webService.getMoviePopular(page).enqueue(new Callback<Movie>() {
            @Override
            public void onResponse(@NonNull Call<Movie> call, @NonNull Response<Movie> response) {
                movieSingleLiveEvent.setValue(Resource.success(Objects.requireNonNull(response.body())));
            }

            @Override
            public void onFailure(@NonNull Call<Movie> call, @NonNull Throwable t) {
                movieSingleLiveEvent.setValue(Resource.error(t.getMessage(), null));
            }
        });
    }

    public void getMovieTop(SingleLiveEvent<Resource<Movie>> movieTopSingleLiveEvent, Integer page) {
        webService.getMovieTopRated(page).enqueue(new Callback<Movie>() {
            @Override
            public void onResponse(@NonNull Call<Movie> call, @NonNull Response<Movie> response) {
                movieTopSingleLiveEvent.setValue(Resource.success(Objects.requireNonNull(response.body())));
            }

            @Override
            public void onFailure(@NonNull Call<Movie> call, @NonNull Throwable t) {
                movieTopSingleLiveEvent.setValue(Resource.error(t.getMessage(), null));
            }
        });
    }
}

File: main/src/main/java/com/example/main/data/source/remote/MainRepository.kt
package com.example.main.data.source.remote


import com.example.core.data.model.Resource
import com.example.core.data.model.movie.Movie
import com.example.core.util.SingleLiveEvent

import java.util.Objects

import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class MainRepository : DataSourceMainRepository {

    private val webService = ApiUtilsMain().listMovieService

    override fun getMovies(movieSingleLiveEvent: SingleLiveEvent<Resource<Movie>>, page: Int) {
        webService.getMoviePopular(page).enqueue(object : Callback<Movie> {
            override fun onResponse(call: Call<Movie>, response: Response<Movie>) {
                movieSingleLiveEvent.value =
                    Resource.success(Objects.requireNonNull<Movie>(response.body()))
            }

            override fun onFailure(call: Call<Movie>, t: Throwable) {
                movieSingleLiveEvent.value = Resource.error(t.message, null)
            }
        })
    }

    override fun getTopMovies(
        movieTopSingleLiveEvent: SingleLiveEvent<Resource<Movie>>,
        page: Int
    ) {
        webService.getMovieTopRated(page).enqueue(object : Callback<Movie> {
            override fun onResponse(call: Call<Movie>, response: Response<Movie>) {
                movieTopSingleLiveEvent.value =
                    Resource.success(Objects.requireNonNull<Movie>(response.body()))
            }

            override fun onFailure(call: Call<Movie>, t: Throwable) {
                movieTopSingleLiveEvent.value = Resource.error(t.message, null)
            }
        })
    }
}
